src/main/java/seedu/address/model/person/Role.java

```diff
@@ -0,0 +1,152 @@
+package seedu.address.model.person;
+
+import static java.util.Objects.requireNonNull;
+import static seedu.address.commons.util.AppUtil.checkArgument;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Represents the roles taken by a person in NUS.
+ */
+public class Role {
+    public static final String MESSAGE_CONSTRAINTS = "Role consists of two parts, module code and role type.\n"
+            + "Module code can take on any values, and should not be blank.\n"
+            + "Role type can be blank, and or any of the following values(case insensitive) for each role type:\n"
+            + "1. Student: Student or leave blank\n"
+            + "2. Tutor: Tutor or TA\n"
+            + "3. Professor: Professor or Prof\n";
+    public static final String VALIDATION_REGEX = "^$|student|tutor|ta|professor|prof";
+    private final HashMap<ModuleCode, RoleType> roles;
+
+    /**
+     * Default constructor for a {@code Role}.
+     * This is for creating roles for a Person based on the module code and role types
+     * extracted from an add command.
+     *
+     * @param moduleCodes Array of modules codes
+     * @param roleTypes Array of role type String
+     */
+    public Role(ModuleCode[] moduleCodes, String[] roleTypes) {
+        requireNonNull(moduleCodes);
```

Comment:
hmm do you want a constructor of `Role(ModuleCode[] moduleCodes, RoleType[] roleTypes)` instead? assume that this class will receive the `RoleType[]`, then the conversion of user input of type `String` to the `RoleType` will be done in the parser

Reply:
Thanks for the suggestion! I realized that all other classes that represents attribute of a Person has validation steps at the parsing command, object creation and loading JSON data stage. For the Role class, since the data passed in during parsing phase and the actual data stored are of different types, I think if we are to implement validations are all three stages we need multiple validation methods to handle different types of data. My thoughts are like these:
1. For parsing stage, the parser just checks if the module code and the roles keywords are valid inputs and then convert them to respective ModuleCode[] and RoleType[] array and pass these two array to the constrcutor
2. For constructor itself, we can validate the inputs by checking if there's any null object in the given array
3. For `JsonAdaptedRole`(should be implemented for save/load feature), we also check if there's any null objects and then convert them to their respective Model type.
Do you think this approach is sufficient to make sure that the program won't break due to any invalid input or messing up of the storage? Thank you for your time

Reply:
yep I agree with your thoughts! so just to paraphrase, there should be a

- `String -> RoleType` convert + validity check
- `RoleType` non-null check
- `ModuleCode` non-null check

1. Parsing stage uses all 3 (similar to `Tag`)
2. Constructor uses non-null checkers only because parsing stage will return `RoleType[]`, so it's not possible to check the `String` anymore. I'm not sure why the provided `Tag` class runs its validity checker both at the parser and the constructor
3. `JsonAdaptedRole` should use all 3, since it is basically doing parsing and constructing. However, it seems like `JsonAdaptedTag` does not check for `null` at all, which is a bug in AB3